src/Analyser/MutatingScope.php

```diff
@@ -294,6 +294,54 @@ public function enterDeclareStrictTypes(): self
 		);
 	}
 
+	public function rememberConstructorScope(): self
+	{
+		$expressionTypes = [];
+		foreach ($this->expressionTypes as $exprString => $expressionTypeHolder) {
+			$expr = $expressionTypeHolder->getExpr();
+			if (!$expr instanceof ConstFetch) {
+				continue;
+			}
+			$expressionTypes[$exprString] = $expressionTypeHolder;
+		}
+
+		$nativeExpressionTypes = [];
+		foreach ($this->nativeExpressionTypes as $exprString => $expressionTypeHolder) {
+			$expr = $expressionTypeHolder->getExpr();
+			if (!$expr instanceof ConstFetch) {
+				continue;
+			}
+
+			$nativeExpressionTypes[$exprString] = $expressionTypeHolder;
```

Comment:
I'd say you should go through these expressions and exclude everything that references variables (using NodeFinder), except for `$this`. That way you'd be able to narrow property types as well.

Reply:
Also: this isn't the only required change in MutatingScope. You need to make sure these things are re-used in enterClassMethod.

Reply:
Oh right, it'd work only for readonly properties.

Reply:
do you think it should still be implemented?

or is it now worth the effort because in the readonly case its likely combined with property promotion and this is likely already narrowed with a `@param` type

Reply:
I remember some cases where it was also used for `class_exists`, to remove false positives from the rest of the class. So I'd say this might be useful and it's not as much effort.

Oh: we need to use the original scope (not the narrowed one) for static methods.

Reply:
Also the narrowed scope is useful for property hook bodies so please test that (on 8.4+).

Reply:
> Oh: we need to use the original scope (not the narrowed one) for static methods.

yeah, thats the reason why I sorted static methods to be analyzed first.. that way they work with a clean scope